Create comparison section (#255)
waku-org/waku.guide: docs/learn/compare/libp2p.md

---
title: Waku vs Libp2p
hide_table_of_contents: true
displayed_sidebar: learn
---

# Waku vs libp2p

## Core Distinction

**libp2p** is a general-purpose P2P networking framework providing transport, discovery, and pub/sub primitives for any decentralized application.

**Waku** is a privacy-preserving messaging protocol built on top of libp2p, extending it with metadata privacy and zero-knowledge DoS protection.

These protocols are complementary, Waku inherits libp2p's networking foundation and adds privacy-first messaging capabilities.

## Quick Comparison

| Aspect | libp2p | Waku |
|--------|--------|------|

> | **Purpose** | General P2P infrastructure | Privacy-preserving messaging |

Reviewer: ```suggestion
| **Purpose** | General P2P libraries | Privacy-preserving, anonymouse, censorship-resistant messaging SDKs and infrastructure. |
```

| **Metadata Privacy** | None (connection patterns visible) | Strong (no-signature relay, k-anonymity) |

> | **DoS Protection** | Peer scoring only | RLN (zero-knowledge proofs) |

Reviewer: ```suggestion
| **DoS Protection** | Peer scoring, no bandwidth capping, message inspection needed for validation | RLN (anonymous permissionless bandwidth capping) |
```

| **Latency** | ~10-50ms overhead | ~500ms average |
| **Message Size** | Transport-limited | 150KB max (4KB recommended) |

> | **Offline Delivery** | Application layer required | Built-in Store protocol (12+ hours) |

Reviewer: ```suggestion
| **Offline Delivery** | Application layer required | Built-in caching protocol (store) and repair (SDS-repair) |
```

| **Production Scale** | ETH2 (100K+ validators), IPFS | Status (200K users), RAILGUN, Waku Network (80K capacity) |
| **Maturity** | 6+ years, battle-tested | 4+ years (v2 since 2021), production-ready |
| **Implementations** | Go, Rust, JS, 10+ languages | Nim (reference), Go, JS, Rust (RLN) |

## Privacy Architecture

### libp2p Privacy Model
- Strong transport encryption (TLS 1.3, Noise)
- Forward secrecy through key rotation
- **No metadata protection**: DHT queries, connection patterns, and relay routing are visible
- GossipSub messages signed by default, exposing sender identity
- Requires application-layer privacy additions

### Waku Privacy Model
- Same transport encryption as libp2p
- **Strong metadata privacy**: no-signature policy prevents sender identification
- Content topic strategies provide k-anonymity
- **Formal privacy guarantees**: proven sender anonymity against single-node adversaries

> - **Zero-knowledge DoS protection**: Rate Limiting Nullifiers (RLN) prevent spam without revealing identity

Reviewer: ```suggestion
- **Bandwidth capping**: Rate Limiting Nullifiers (RLN) prevent network flood in an anonymous and permissionless manner (smart contract gates memberships)
```
